cluster/router/condition/matcher/base/matcher.go

```diff
@@ -0,0 +1,135 @@
+	"dubbo.apache.org/dubbo-go/v3/common/constant"
+	"dubbo.apache.org/dubbo-go/v3/common/extension"
+	"dubbo.apache.org/dubbo-go/v3/protocol"
+)
+
+// ConditionMatcher records the match and mismatch patterns of this matcher while at the same time
+// provides the common match logics.
+type ConditionMatcher struct {
+	Matches       map[string]struct{}
+	Mismatches    map[string]struct{}
+	ValueMatchers []pattern.ValuePattern
+
+	Matcher matcher.ConditionMatcher
+	Key     string
+}
+
+func (c *ConditionMatcher) GetValue(sample map[string]string, url *common.URL, invocation protocol.Invocation) string {
+	return ""
+}
+
+func (c *ConditionMatcher) GetSampleValueFromUrl(conditionKey string, sample map[string]string, param *common.URL, invocation protocol.Invocation) string {
+	var sampleValue string
+	//get real invoked method name from invocation
+	if invocation != nil && (constant.MethodKey == conditionKey || constant.MethodsKey == conditionKey) {
+		sampleValue = invocation.MethodName()
+	} else {
+		sampleValue = sample[conditionKey]
+	}
+	return sampleValue
+}
+
+func (c *ConditionMatcher) IsMatch(sample map[string]string, param *common.URL, invocation protocol.Invocation, isWhenCondition bool) bool {
+	value := c.Matcher.GetValue(sample, param, invocation)
+	if value == "" {
+		// if key does not present in whichever of url, invocation or attachment based on the matcher type, then return false.
+		return false
+	}
+
+	if len(c.Matches) != 0 && len(c.Mismatches) == 0 {
+		for match, _ := range c.Matches {
+			if c.DoPatternMatch(match, value, param, invocation, isWhenCondition) {
+				return true
+			}
+		}
+		return false
+	}
+
+	if len(c.Mismatches) != 0 && len(c.Matches) == 0 {
+		for mismatch, _ := range c.Mismatches {
+			if c.DoPatternMatch(mismatch, value, param, invocation, isWhenCondition) {
+				return false
+			}
+		}
+		return true
+	}
+
+	if len(c.Matches) != 0 && len(c.Mismatches) != 0 {
+		//when both mismatches and matches contain the same value, then using mismatches first
+		for mismatch, _ := range c.Mismatches {
+			if c.DoPatternMatch(mismatch, value, param, invocation, isWhenCondition) {
+				return false
+			}
+		}
+
+		for match, _ := range c.Matches {
+			if c.DoPatternMatch(match, value, param, invocation, isWhenCondition) {
+				return true
+			}
+		}
+		return false
+	}
+	return false
+}
+
+func (c *ConditionMatcher) GetMatches() map[string]struct{} {
+	return c.Matches
+}
+
+func (c *ConditionMatcher) GetMismatches() map[string]struct{} {
+	return c.Mismatches
+}
+
+func (c *ConditionMatcher) DoPatternMatch(pattern string, value string, url *common.URL, invocation protocol.Invocation, isWhenCondition bool) bool {
+	for _, valueMatcher := range c.ValueMatchers {
+		if valueMatcher.ShouldMatch(pattern) {
+			return valueMatcher.Match(pattern, value, url, invocation, isWhenCondition)
+		}
+	}
+	// this should never happen.
+	logger.Error("Executing condition rule value match expression error.")
```

Comment:
Why do you say: "this should never happen"? Is this because the wildcard must do match once?

Reply:
> Why do you say: "this should never happen"? Is this because the wildcard must do match once?
There should at least has one ValueMatcher instance that applies to all patterns, will force to use wildcard matcher now.

Reply:
So I think the comment should be "If no value matcher is available, then using a wildcard value matcher".

Reply:
thx, I will modify